docs: hll&udd (#1844)
GreptimeTeam/docs: docs/reference/sql/functions/approximate.md

:::

## Approximate Count Distinct (HLL)

The `hll` function is used to calculate the approximate count distinct of a set of values. It uses [HyperLogLog](https://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf) (HLL) algorithm for efficient memory usage and speed. Three functions are provided for this purpose:
- `hll(value)` to create a HyperLogLog state in binary from a given column.
- `hll_merge(hll_state)` to merge multiple HyperLogLog states into one.
- `hll_count(hll_state)` to get the approximate count distinct from a HyperLogLog state.

Notice that due to the approximate nature of the algorithm, the results may not be exact but are usually very close to the actual count distinct, and the larger the dataset, the more accurate the results will be. The relative error of the HyperLogLog algorithm is about 1.04/sqrt(m), where m is the number of registers used in the algorithm. GreptimeDB uses 16384 registers by default, which gives a relative error of about 0.008125(or 0.8125%).

### Usage Example
This example demonstrates how to use the `hll` functions to calculate the approximate count distinct of user visits.

> ```sql

Author: > please split this large SQL example to several little ones.

splited, but still want to also keep the full usage example tooo

-- results as follows:
-- +------------+---------------------+
-- | url        | time_window         |
-- +------------+---------------------+
-- | /dashboard | 2025-03-04 00:00:00 |
-- | /dashboard | 2025-03-04 00:00:10 |
-- | /not_found | 2025-03-04 00:00:10 |
-- +------------+---------------------+

-- in addition, we can aggregate the 10-second data to a 1-minute level by merging the HyperLogLog states using `hll_merge`.
SELECT
    `url`,
    date_bin('1 minute' :: INTERVAL, `time_window`) AS time_window_1m,
    hll_count(hll_merge(state)) as uv_per_min
FROM
    access_log_10s
GROUP BY
    `url`,
    date_bin('1 minute' :: INTERVAL, `time_window`);

-- results as follows:
-- +------------+---------------------+------------+
-- | url        | time_window_1m      | uv_per_min |
-- +------------+---------------------+------------+
-- | /dashboard | 2025-03-04 00:00:00 |          3 |
-- | /not_found | 2025-03-04 00:00:00 |          1 |
-- +------------+---------------------+------------+
```

## Approximate Quantile (UDDSketch)

Three functions are provided for approximate quantile calculation using the [UDDSketch](https://arxiv.org/abs/2004.08604) algorithm:

> - `uddsketch(bucket_num, error_rate, value)` to create a UDDSketch state in binary from a given column, the `bucket_num` is the number of buckets to use for the sketch, and `error_rate` is the desired error rate for the quantile calculation.

Author: > please explain more about the arguments. how to determine the `bucket_num`, `error_rate` and `value`.

explained in the Caveats section

---
keywords: [Approximate functions, approximate count distinct, approximate quantile, SQL functions]
description: Lists and describes approximate functions available in GreptimeDB, including their usage and examples.
---

# Approximate Functions

This page lists approximate functions in GreptimeDB, which are used for approximate data analysis.

:::warning
The following approximate functions is currently experimental and may change in future releases.
:::

## Approximate Count Distinct (HLL)

> The `hll` function is used to calculate the approximate count distinct of a set of values. It uses [HyperLogLog](https://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf) (HLL) algorithm for efficient memory usage and speed. Three functions are provided for this purpose, described in following chapters:

Reviewer: ```suggestion
The HyperLogLog (HLL) algorithm is used to calculate the approximate count distinct of a set of values. It provides efficient memory usage and speed for this purpose. Three functions are provided to work with the HLL algorithm, described in following chapters:
```

I think this description is clearer because `hll` is just one function in the family implementing this algorithm, not all of them; emphasizing it alone can easily cause misunderstanding.

> The `hll` function is used to calculate the approximate count distinct of a set of values. It uses [HyperLogLog](https://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf) (HLL) algorithm for efficient memory usage and speed. Three functions are provided for this purpose, described in following chapters:

Reviewer: ```suggestion
The HyperLogLog (HLL) algorithm is used to calculate the approximate count distinct of a set of values. It provides efficient memory usage and speed for this purpose. Three functions are provided to work with the HLL algorithm, described in following chapters:
```

I think this description is clearer because `hll` is just one function in the family implementing this algorithm, not all of them; emphasizing it alone can easily cause misunderstanding.

### `hll_merge`

`hll_merge(hll_state)` merges multiple HyperLogLog states into one. This is useful when you want to combine the results of multiple HLL calculations, such as when aggregating data from different time windows or sources. The `hll_state` parameter is the binary representation of the HLL state created by [`hll`](#hll). The merged state can then be used to calculate the approximate count distinct across all the merged states. See [Full Usage Example](#Full_Usage_Example) for a full example of how to use this function in combination with other functions to calculate approximate count distinct.

### `hll_count`

`hll_count(hll_state)` retrieves the approximate count distinct from a HyperLogLog state. This function takes the HLL state created by `hll` or merged by `hll_merge` and returns the approximate count of distinct values. See [Full Usage Example](#Full_Usage_Example) for a full example of how to use this function in combination with other functions to calculate approximate count distinct.

> ### Full Usage Example

Reviewer: ```suggestion
### HLL Functions Flow

```
10:00:00-10:00:10    10:00:10-10:00:20    10:00:20-10:00:30
┌─────────────────┐   ┌─────────────────┐   ┌─────────────────┐
│user_id: 1,1,2   │   │user_id: 2,2,3   │   │user_id: 1,3,4   │
│(3 records)      │   │(3 records)      │   │(3 records)      │
└─────────────────┘   └─────────────────┘   └─────────────────┘
         │                       │                       │
         ▼                       ▼                       ▼
     ┌────────┐              ┌────────┐              ┌────────┐
     │HLL 1   │              │HLL 2   │              │HLL 3   │
     │(Binary)│              │(Binary)│              │(Binary)│
     └────────┘              └────────┘              └────────┘
         │                       │                       │
         ▼                       ▼                       ▼
     ┌────────┐              ┌────────┐              ┌────────┐
     │hll_count│             │hll_count│             │hll_count│
     │   ~2   │              │   ~2   │              │   ~3   │
     └────────┘              └────────┘              └────────┘
         │                       │                       │
         └───────────┬───────────┴───────────┬───────────┘
                     │                       │
                     ▼                       ▼
                ┌─────────────┐         ┌─────────────┐
                │ hll_merge() │         │ hll_count() │
                └─────────────┘         └─────────────┘
                     │                       │
                     ▼                       ▼
                ┌─────────────┐         ┌─────────────┐
                │Merged HLL   │         │   ~4        │
                │(Binary)     │         │(approx)     │
                └─────────────┘         └─────────────┘
                     │                       │
                     ▼                       ▼
                ┌─────────────┐         ┌─────────────┐
                │10:00:00-    │         │Total unique │
                │10:00:30     │         │users: ~4    │
                │(30 seconds) │         │             │
                └─────────────┘         └─────────────┘
```
```

Is this helpful?

Author: will convert this to a svg and add one for uddsketch too, also will add explain text

### How to determine `bucket_num` and `error_rate`

The `bucket_num` parameter sets the maximum number of internal containers the sketch can use, directly controlling its memory footprint. Think of it as the physical storage capacity for tracking different value ranges. A larger `bucket_num` allows the sketch to accurately represent a wider dynamic range of data (i.e., a larger ratio between the maximum and minimum values). If this limit is too small for your data, the sketch will be forced to merge very high or low values, which degrades its accuracy. A recommended value for `bucket_num` is 128, which provides a good balance between accuracy and memory usage for most use cases. 

The `error_rate` defines the desired precision for your quantile calculations. It guarantees that any computed quantile (e.g., p99) is within a certain *relative* percentage of the true value. For example, an `error_rate` of `0.01` ensures the result is within 1% of the actual value. A smaller `error_rate` provides higher accuracy, as it forces the sketch to use more granular buckets to distinguish between closer numbers.

> These two parameters create a direct trade-off. To achieve the high precision promised by a small `error_rate`, the sketch needs a sufficient `bucket_num`, especially for data that spans a wide range. `bucket_num` acts as the physical limit on accuracy. If your `bucket_num` is restricted by memory constraints, setting the `error_rate` to an extremely small value will not improve precision beyond the limit imposed by the available buckets.

Reviewer 2: I'm not sure if my understanding is correct:

```suggestion
These two parameters create a direct trade-off. To achieve the high precision promised by a small `error_rate`, the sketch needs a sufficient `bucket_num`, especially for data that spans a wide range. `bucket_num` acts as the physical limit on accuracy. If your `bucket_num` is restricted by memory constraints, setting the `error_rate` to an extremely small value will not improve precision because the limit imposed by the available buckets.
```

-- | /not_found | 2025-03-04 00:00:10 |                               1 |
-- +------------+---------------------+---------------------------------+

-- in addition, we can aggregate the 10-second data to a 1-minute level by merging the HyperLogLog states using `hll_merge`.
SELECT
    `url`,
    date_bin('1 minute' :: INTERVAL, `time_window`) AS time_window_1m,
    hll_count(hll_merge(state)) as uv_per_min
FROM
    access_log_10s
GROUP BY
    `url`,
    date_bin('1 minute' :: INTERVAL, `time_window`);

-- results as follows:
-- +------------+---------------------+------------+
-- | url        | time_window_1m      | uv_per_min |
-- +------------+---------------------+------------+
-- | /dashboard | 2025-03-04 00:00:00 |          3 |
-- | /not_found | 2025-03-04 00:00:00 |          1 |
-- +------------+---------------------+------------+
```

## Approximate Quantile (UDDSketch)

Three functions are provided for approximate quantile calculation using the [UDDSketch](https://arxiv.org/abs/2004.08604) algorithm.

### `uddsketch_state`

The `uddsketch_state` function is used to create a UDDSketch state in binary from a given column. It takes three parameters: 

> - `bucket_num`, which is the number of buckets to use for the sketch, 

Reviewer 2: please add a link to `how to determine bucket_num value`

```sql
-- calculate the approximate 99th percentile from the UDDSketch state in a 5-second windowed query
SELECT
    time_window,
    uddsketch_calc(0.99, `percentile_state`) AS p99
FROM
    percentile_5s;

-- results as follows:
-- +---------------------+--------------------+
-- | time_window         | p99                |
-- +---------------------+--------------------+
-- | 1970-01-01 00:00:00 |  40.04777053326359 |
-- | 1970-01-01 00:00:05 |  89.13032933635911 |
-- | 1970-01-01 00:00:10 | 100.49456770856492 |
-- +---------------------+--------------------+

-- calculate the approximate 99th percentile from the merged UDDSketch state in a 1-minute windowed query
SELECT
    date_bin('1 minute' :: INTERVAL, `time_window`) AS time_window_1m,
    uddsketch_calc(0.99, uddsketch_merge(128, 0.01, `percentile_state`)) AS p99
FROM
    percentile_5s
GROUP BY
    time_window_1m;

-- results as follows:
-- +---------------------+--------------------+
-- | time_window_1m      | p99                |
-- +---------------------+--------------------+
-- | 1970-01-01 00:00:00 | 100.49456770856492 |
-- +---------------------+--------------------+
```

> ### Caveats

Reviewer 2: ```suggestion
### How to determine `bucket_num` value
```
